src/native/libs/System.Globalization.Native/pal_calendarData.m

```diff
@@ -141,4 +170,121 @@
+            NSDateComponents *eraComponents = [japaneseCalendar components:NSCalendarUnitEra fromDate:date];
+            currentEra = [eraComponents era];
+            if (currentEra == era)
+            {
+                for (int day = 0; day < 31; day++)
+                {
+                    // subtract 1 day at a time until we get out of the specified Era
+                    startDateComponents.day = startDateComponents.day - 1;
+                    date = [japaneseCalendar dateFromComponents:startDateComponents];
+                    eraComponents = [japaneseCalendar components:NSCalendarUnitEra fromDate:date];
+                    currentEra = [eraComponents era];
+                    if (currentEra != era)
+                    {
+                        // add back 1 day to get back into the specified Era
+                        startDateComponents.day = startDateComponents.day + 1;
+                        date = [japaneseCalendar dateFromComponents:startDateComponents];
+                        NSCalendar *gregorianCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
+                        NSDateComponents *components = [gregorianCalendar components:NSCalendarUnitDay | NSCalendarUnitMonth | NSCalendarUnitYear fromDate:date];
+                        *startYear = [components year];
+                        *startMonth = [components month];
+                        *startDay = [components day];
+                        return 1;
+                    }
+                }
+            }
+            // add 1 month at a time until we get into the specified Era
+            startDateComponents.month = startDateComponents.month + 1;
+            date = [japaneseCalendar dateFromComponents:startDateComponents];
+            eraComponents = [japaneseCalendar components:NSCalendarUnitEra fromDate:date];
+            currentEra = [eraComponents era];
+        }
+
+        return 0;
+    }
+}
+
+/*
+Function:
+GetCalendarsNative
+
+Returns the list of CalendarIds that are available for the specified locale.
+*/
+int32_t GlobalizationNative_GetCalendarsNative(CalendarId* calendars, int32_t calendarsCapacity)
+{
+    @autoreleasepool
+    {
+        NSArray *calendarIdentifiers = @[
+            NSCalendarIdentifierGregorian,
+            NSCalendarIdentifierBuddhist,
+            NSCalendarIdentifierChinese,
+            NSCalendarIdentifierCoptic,
+            NSCalendarIdentifierEthiopicAmeteMihret,
+            NSCalendarIdentifierEthiopicAmeteAlem,
+            NSCalendarIdentifierHebrew,
+            NSCalendarIdentifierISO8601,
+            NSCalendarIdentifierIndian,
+            NSCalendarIdentifierIslamicCivil,
+            NSCalendarIdentifierIslamicTabular,
+            NSCalendarIdentifierIslamicUmmAlQura,
+            NSCalendarIdentifierIslamic,
+            NSCalendarIdentifierJapanese,
+            NSCalendarIdentifierPersian,
+            NSCalendarIdentifierRepublicOfChina,
+        ];
```

Comment:
It seems like it would be problematic if certain regions explicitly exclude particular identifiers, yet we provide them all by default.
Would it be tedious to find out what the ICU returns for different locales? `ucal_getKeywordValuesForLocale`
Are these calendar identifiers all available regardless of region when developing with `NSCalendar`?

Reply:
I'm investigating more connection between `NSLocale` and `NSCalendar`. So far my understanding is `NSLocale` doesn't have info about supported calendars, and for `NSCalendar` is possible to set locale but didn't see any restrictions there.

Reply:
Looks like for `ICU` it takes info from here https://github.com/dotnet/icu/blob/dotnet/main/icu/icu4c/source/data/misc/supplementalData.txt#L1888 . In case of native there is no preferred calendar values for specific locale, and no restriction to set `NSCalendar`'s locale ( for example it is possible to have `NSCalendarIdentifierJapanese` with `he_IL` (Hebrew) locale). Considering this, it should be okay to return for all locales all possible calendars. 
The only behavioural change is that with this implementation  default calendar will always be `gregorian`. I will add that in the documantation.

Reply:
Are there any tests that depend on this info?
Have any of the public APIs that depend on this function changed in behavior because we are returning a full set regardless of locale? For example, https://learn.microsoft.com/en-us/dotnet/api/system.globalization.cultureinfo.optionalcalendars?view=net-7.0 it seems like we should only have 6 calendars for "ar-SA" culture. We should make sure that is still maintained because that is documented.

Reply:
- I checked what ICU is returning now for "ar-SA". 

  ```
  The ar-SA culture supports the following calendars:
  System.Globalization.UmAlQuraCalendar
  System.Globalization.GregorianCalendar , callType = Localized
  System.Globalization.HijriCalendar
  ```
  which is not the same as mentioned in the link you shared (So either doc is not updated or on mobile it has different output)

- Hybrid mode on other hand will return all calendars as supported.

@steveisok @akoeplinger could you please take a look if it is acceptable to have these behavioural changes on hybrid mode? 
With this implementation on hybrid mode `OptionalCalendars` will return all calendars and default calendar will be `Gregorian`.

Reply:
I think returning the additional calendars in `OptionalCalendars` is not a big issue, as you said we already seem to be deviating from the documented behavior on desktop/ICU.

But the default calendar needs to be correct, e.g. on `ar-SA` it needs to return `UmAlQuraCalendar`.
You can get the calendar associated with an NSLocale via the `calendarIdentifier`, e.g. like this:
```objc
    NSLocale *currentLocale = [[NSLocale alloc] initWithLocaleIdentifier:@"ar-SA"];
    printf("%s", [[currentLocale calendarIdentifier] UTF8String]);

   // this prints "islamic-umalqura"
```

We can use that to sort the corresponding ID into the first spot of the array (which decides the default calendar).

Reply:
sounds good, will update default calendar.